src/gsAssembler/gsPatchRule.h

```diff
@@ -0,0 +1,267 @@
+
+    /**
+     * @brief      Computes the number of quadrature points to exactly integrate the space
+     *
+     * @param[in]  knots  knot vector
+     *
+     * @return     Integer number of quadrature points
+     */
+    index_t _numQuads(const gsKnotVector<T> knots) const
+    {
+        index_t ndof = (m_deg + 1)*knots.numElements() - (m_reg+1)*(knots.numElements()-1);
+        ndof = std::ceil(ndof/2.0);
+        return ndof;
+    }
+
+    /**
+     * @brief      Determines whether the specified knot vector is symmetric.
+     *
+     * @param[in]  knots  knot vector
+     *
+     * @return     True if the specified knots is symmetric, False otherwise.
+     */
+    bool _isSymmetric(const gsKnotVector<T> knots) const
+    {
+        gsKnotVector<T> copy = knots;
+        copy.reverse();
+        std::vector<T> result;
+        knots.symDifference(copy,result);
+        return result.size()==0 ? true : false;
+    }
+
+
+    /**
+     * @brief      Initializes the knot vector for the patch-rule implementation
+     *
+     * The knot vector for the basis in a specific direction is modified:
+     * 1) the number of knots is made even. If this is not the case, an extra knot is added in the middle
+     * 2) over-integration is applied. In this case, p knots are added to the boundary elements
+     *
+     * @param[in]  Bbasis  a gsBSplineBasis
+     *
+     * @return     knot vector
+     */
+    gsKnotVector<T> _init(const gsBSplineBasis<T> * Bbasis) const;
+
+    /**
+     * @brief      Integrates all basis functions from a knot vector (numerically with Gauss)
+     *
+     * This computes the integrals of the basis function for the newton integration.
+     *
+     * @param      knots  The knots
+     *
+     * @return     Greville points and Exact integrals of the basis
+     */
+    std::pair<gsMatrix<T>,gsMatrix<T> > _integrate(const gsKnotVector<T> & knots ) const;
+
+    /**
+     * @brief      Computes the points and weights for the patch rule using Newton iterations
+     *
+     * @param      knots      The knots of the basis
+     * @param      greville   The greville point
+     * @param      integrals  The exact integrals
+     * @param[in]  tol        The tolerance
+     *
+     * @return     knots and weights
+     */
+    std::pair<gsVector<T>,gsVector<T> > _compute(const gsKnotVector<T> & knots,
+                                                const gsMatrix<T> & greville,
+                                                const gsVector<T> & integrals,
+                                                const T tol = 1e-10) const;
+
+private:
+    const gsBasis<T> * m_basis;
+    const index_t m_deg,m_reg;
+    const bool m_over;
+    const short_t m_fixDir;
+    mutable index_t m_nQuad;
+    mutable bool m_symmetry;
+
+    std::vector<gsVector<T> > m_nodes;
+    std::vector<gsVector<T> > m_weights;
+
+    mutable gsVector<T> m_end;
+
+    mutable typename gsSparseSolver<T>::QR m_solver;
+
+    mutable size_t m_dim;
```

Comment:
this may not be mutable, probably also some of the others, please check (only make something mutable after you get an error that modification was forbitten somewhere)

Reply:
Only the one for the solver was needed